### DISCUSSION TITLE
API connection can't be established with ollama running on an Azure VM

### DISCUSSION THREAD

**Discussion by pwab** [open-webui/open-webui#434] [Q&A] [answered]
Hi 👋 

I'm trying to build a test environment like this:

- Azure
  - Ubuntu 22.04.3 LTS VM with running ollama (0.1.18)
  - Port mapping to port 80
  - URL to get "ollama is running": `http://xxx.yyy.cloudapp.azure.com`
- Local Computer
  - Podman with docker container running ollama-webui (v1.0.0-alpha.61) with
    - `podman run -d --network=host -p 3000:8080 -e OLLAMA_API_BASE_URL=http://xxx.yyy.cloudapp.azure.com/api -v ollama-webui-extern:/app/backend/data --name ollama-webui-extern --restart always ghcr.io/ollama-webui/ollama-webui:main`
  - For testing purposes: A local ollama instance also running with podman on `localhost:11434`

For a quick test I'm using the following ollama API request:
`curl $URL/api/generate -d '{"model": "zephyr","prompt": "Why is the sky blue?"}'`

These are my results:
- Local ollama instance
  - test request with curl ✅ 
  - use for ollama-webui ✅ 
- Azure instance
  - test request with curl ✅
  - use for ollama-webui ⛔ 

When trying to access the Azure instance I'm getting the notification `Server connection failed` (and not `Ollama WebUI: Server Connection Error` which is promped when using a wrong port). Also the browsers console is logging the following:
```
SyntaxError: Unexpected end of JSON input
    at index.d890aff4.js:1:1070
    at async h (index.d890aff4.js:1:867)
    at async 2.03a95dfe.js:35:2083
```

The container itself is able to reach the API endpoint. I confirmed that by logging into the container, installing curl and running my test request from there. So there shouldn't be a problem with the network here I guess.

Could there be a problem when trying to access an external ollama which runs on port 80? Or maybe the URL pattern is a problem? I have no clue what else could be wrong here.

**Comment by justinh-rahb** [comment, marked as answer]
Have you passed the OLLAMA_ORIGINS environment variable on the azure host and set it to the URL your webui will use (or *)?

Also, you can't use `--networking host` and `-p` port mapping at the same time, or at any rate there's no point in doing so. Your ollama instance would then be accessible from port 8080 if you're using host networking.

**Reply by pwab** [reply]
I only used `--networking host` additionally because it was mentioned in the troubleshooting docs. There was a little hint that using `-p` isn't neccessary when doing so but it wouldn't have an effect. I'll change my podman command. Thanks.

I wasn't aware of the [OLLAMA_ORIGINS](https://github.com/jmorganca/ollama/blob/6164f378f2780f8563f08e8cba315450f6b3b94e/docs/faq.md?plain=1#L50C16-L50C16) env variable. Maybe because I only read the ollama-webui docs and not the ollama docs from start to finish.
It seems to be promising. But as I'm able to reach the ollama API with curl does it make any difference for ollama-webui?

I'll try to test soon and give feedback.

**Reply by justinh-rahb** [reply]
Sure does. A curl command will work, apps would block the request because it fails CORS check.

ETA: The host networking thing is really only applicable to issues related to accessing the ollama API running on the same host as the container. Since you're using external API anyway it shouldn't matter.

**Reply by pwab** [reply]
Thanks for clarification. CORS is a topic that I should really look into in more detail.
Just if anyone else is interested: I tried the following [step by step guide](https://github.com/jmorganca/ollama/issues/1157#issuecomment-1815550029) which is also documented in the [FAQ](https://github.com/jmorganca/ollama/blob/main/docs/faq.md#how-do-i-use-ollama-server-environment-variables-on-linux) in a more comprehensive way.

---

This is a bit stupid, but I just remembered that I used caddy as a reverse proxy for the port mapping from 11434 to 80 as a test. Because I'm pretty sure that this also plays a role here, but it no longer has any relation to the original question, the issue is closed for me for now.

I will now rebuild my system a bit and run ollama-webui on the Azure VM together with ollama. That will make things easier. Thanks for your efforts!

**Reply by justinh-rahb** [reply]
Yes, it is likely the proxy in between your ollama and ollama-webui causing the issue here. I'd recommend trying to connect them as directly as you can, or use some sort of tunnelling if you're worried about unauthorized access (you probably should be if running it on a public cloud).
